### PR TITLE
add ndarray typedef

### PR DESCRIPTION
**Pull request by tamiko**
TODO:
 - [x] add a test
 - [x] finish up documentation

### CODE REVIEW

**Comment by tamiko** [comment]
@bangerth That's what I would have in mind. Rest of documentation and a test incoming.

File: include/deal.II/base/ndarray.h [hunk 1]
```diff
@@ -0,0 +1,102 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 1921 by the deal.II authors
```
Comment (peterrum) [inline include/deal.II/base/ndarray.h:3]: Are we going back in time ;)

Comment (bangerth) [reply include/deal.II/base/ndarray.h:3]: Ha, that's funny!

File: include/deal.II/base/ndarray.h [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  // == std::array<HelperArray<double, 2, 3, 4>::type, 1>
+  // == std::array<std::array<HelperArray<double, 3, 4>::type, 2>, 1>
+  // == std::array<std::array<std::array<HelperArray<double, 4>::type, 3>, 2>, 1>
+  // == std::array<std::array<std::array<std::array<HelperArray<double>::type, 4>, 3>, 2>, 1>
+  // == std::array<std::array<std::array<std::array<double, 4>, 3>, 2>, 1>
+  //
+  // clang-format on
+  template <typename T, std::size_t... Ns>
+  struct HelperArray
+  {};
```
Comment (bangerth) [inline include/deal.II/base/ndarray.h:44]: I don't think you even need the class body here. A forward declaration would be sufficient and probably lead to clearer error messages.

File: include/deal.II/base/ndarray.h [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+#include <deal.II/base/config.h>
+#include <deal.II/base/exceptions.h>
+
+#include <array>
+
+DEAL_II_NAMESPACE_OPEN
+
+#ifndef DOXYGEN
+namespace
```
Comment (bangerth) [inline include/deal.II/base/ndarray.h:27]: no anonymous namespaces in header files...

File: include/deal.II/base/ndarray.h [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+{
+  // clang-format off
+  //
+  // A variadic template helper class to recursively "unroll" the size
```
Comment (bangerth) [inline include/deal.II/base/ndarray.h:31]: Just make that into a regular doxygen-style comment, even though you specifically tell doxygen to not bother here

File: include/deal.II/base/ndarray.h [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  // == std::array<std::array<std::array<std::array<HelperArray<double>::type, 4>, 3>, 2>, 1>
+  // == std::array<std::array<std::array<std::array<double, 4>, 3>, 2>, 1>
+  //
+  // clang-format on
+  template <typename T, std::size_t... Ns>
+  struct HelperArray
+  {};
+
+  // Recursively define the type typedef of HelperArray<T, N, ...Ns> by
+  // wrapping a std::array around HelperArray<T, Ns...>::type
```
Comment (bangerth) [inline include/deal.II/base/ndarray.h:47]: same here

File: include/deal.II/base/ndarray.h [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+  template <typename T, std::size_t N, std::size_t... Ns>
+  struct HelperArray<T, N, Ns...>
+  {
+    using type = std::array<typename HelperArray<T, Ns...>::type, N>;
+  };
+
+  // End recursion once no std::size_t template parameters are left and
+  // simply set the type typedef to type T
```
Comment (bangerth) [inline include/deal.II/base/ndarray.h:55]: and here

File: include/deal.II/base/ndarray.h [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+  template <typename T>
+  struct HelperArray<T>
+  {
+    using type = T;
+  };
+} // namespace
+#endif // DOXYGEN
+
+/**
+ * A (variadic template) typedef for conveniently defining multidimensional
```
Comment (bangerth) [inline include/deal.II/base/ndarray.h:65]:
```suggestion
 * A (variadic template) `using` declaration for conveniently defining multidimensional
```

File: include/deal.II/base/ndarray.h [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+ * <a href="https://en.cppreference.com/w/cpp/container/array">std::array</a>s
+ *
+ * The problem we try to address with the typedef is the following. Suppose
```
Comment (bangerth) [inline include/deal.II/base/ndarray.h:68]:
```suggestion
 * The problem we try to address is the following. Suppose
```

File: include/deal.II/base/ndarray.h [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+ * you want to create a multdimensional array of doubles of , for example,
```
Comment (bangerth) [inline include/deal.II/base/ndarray.h:69]:
```suggestion
 * you want to create a multdimensional array of doubles of, for example,
```

File: include/deal.II/base/ndarray.h [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+ * rank 3, with sizes 2, 3, 4 for the first, middle, and last index. Then
+ * using C-style arrays you could simply write
+ * @code
+ *   double my_array[2][3][4] = { ... };
+ * @endcode
+ * Nowadays, there are a number of good reasons why using C-style arrays is
+ * usually discouraged (ranging from incompatibilities with STL functions
+ * requiring awkward wrappers, surprises when comparing for equality, etc.)
+ * If you want to do the same, however, using the more modern (and
+ * encouraged) `std::array` then you would have to declare
```
Comment (bangerth) [inline include/deal.II/base/ndarray.h:79]:
```suggestion
 * encouraged) `std::array` class, then you would have to declare
```

File: include/deal.II/base/ndarray.h [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+ * @code
+ *   std::array<std::array<std::array<double, 4>, 3>, 2> = { ... };
+ * @endcode
+ * The repetitions of `std::array` look awkward and, worse, the index
+ * ranges have reversed: the leftmost index has range [0,2), the middle
+ * index has rnage [0,3) and the rightmost index has range [0,4).
+ * We address this issue by providing a typedef ndarray that allows to you
```
Comment (bangerth) [inline include/deal.II/base/ndarray.h:86]:
```suggestion
 * We address this issue by providing a class `ndarray` that allows to you
```
(It's not exactly a class, but it points to a class, so let's not worry too much...

File: include/deal.II/base/ndarray.h [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+ * declare the above stacked std::array type by simply writing:
```
Comment (bangerth) [inline include/deal.II/base/ndarray.h:87]:
```suggestion
 * declare the above stacked `std::array` type by simply writing:
```

File: include/deal.II/base/ndarray.h [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+ * @code
+ *   dealii::ndarray<double, 2, 3, 4> my_array = { ... };
+ * @endcode
+ *
+ * @note dealii::ndarray is merely syntactic sugar in form of a typedef. It
```
Comment (bangerth) [inline include/deal.II/base/ndarray.h:92]:
```suggestion
 * @note dealii::ndarray is merely syntactic sugar in form of a `using` declaration. It
```

File: include/deal.II/base/ndarray.h [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+ * is not a deal.II specific class, but merely a helper to cleanly define
+ * multidimensional arrays realized by "stacked" std::array
```
Comment (bangerth) [inline include/deal.II/base/ndarray.h:94]:
```suggestion
 * multidimensional arrays realized by "stacked" `std::array`.
```

**Review by bangerth — COMMENTED** [review]
Yes, nice. Maybe add a test that checks that these can indeed be initialized via `{...}`?

**Comment by bangerth** [comment]
Also a changelog.

**Comment by bangerth** [comment]
And you need to indent...

File: include/deal.II/base/ndarray.h (hunk 3 above)
Comment (tamiko) [reply include/deal.II/base/ndarray.h:27]: The above use case won't emit any (per compilation unit) symbols - that's why I used an anonymous namespace in the header file.

File: include/deal.II/base/ndarray.h (hunk 1 above)
Comment (tamiko) [reply include/deal.II/base/ndarray.h:3]: https://www.youtube.com/watch?v=2YymGJKhGgY

**Comment by tamiko** [comment]
@bangerth Roger! Will wrap up the documentation and add a test case. Will ping again when this PR is ready.

File: include/deal.II/base/ndarray.h (hunk 1 above)
Comment (tamiko) [reply include/deal.II/base/ndarray.h:3]: https://www.youtube.com/watch?v=m__wmsIn99E

Comment (tamiko) [reply include/deal.II/base/ndarray.h:3]: https://www.youtube.com/watch?v=AM5EYO5wWMA

**Comment by bangerth** [comment]
I kept wondering when @tamiko might get to this, but then remembered that he might not actually have power :-)

**Comment by tamiko** [comment]
@bangerth Internet is back. :partying_face:

File: include/deal.II/base/ndarray.h (hunk 11 above)
Comment (tamiko) [reply include/deal.II/base/ndarray.h:86]: I changed this. "class" is linguistically nicer but what is really going on is that we provide a type alias :-)

**Comment by tamiko** [comment]
```
The following tests FAILED:
    4228 - multigrid/transfer_matrix_free_06.mpirun=15.debug (Timeout)
```

**Review by bangerth — APPROVED** [review]

**Comment by bangerth** [comment]
I've got a follow-up in the works that replaces a lot of uses of nested `std::array`s.

**Pull request merged**